### PR TITLE
Fix float16 log_softmax gradient overflow

### LINKED ISSUE #40409: jax.nn.log_softmax / jax.scipy.special.log_softmax gradient returns NaN on large float16 arrays despite finite forward pass
### Description

`jax.grad` through `log_softmax` on a sufficiently large (~2^16+ element) `float16` array returns an all-`NaN` gradient, even though the forward pass is finite and correct. Reproduces via both `jax.scipy.special.log_softmax` and `jax.nn.log_softmax`, in eager mode and under `jax.jit`.

The true gradient is small, not out of range: for `log_softmax(x).sum()` it is `1 - n * softmax(x)_k`, which for the repro below evaluates to about `[-13.66, 0.98]` (confirmed both via `float32` autodiff and the closed-form expression). `float16`'s max magnitude is 65504, so there's no numerical reason for this to be anything but a small finite value. A same-size, same-dtype *linear* reduction whose true gradient genuinely does exceed `float16`'s range (`x - jnp.sum(x)`) correctly saturates to `-inf` instead — so this isn't range saturation, it's an indeterminate form (`0 * inf` / `inf - inf`) surviving inside the VJP.

This looks related to #13529 at the implementation level (`log_softmax` subtracts a `stop_gradient`'d `jnp.max(x)` before exponentiating), but isn't the same trigger: that issue traces to gradient "smearing" across tied max elements on a deliberately degenerate all-zero-logit input. Forcing this repro's input to have a single, unambiguous, untied max still produces a full `NaN` gradient, and this uses ordinary `N(0,1)` data with no special construction.

## Minimal Reproducer

```python
import jax
import jax.numpy as jnp

x = jax.random.normal(jax.random.PRNGKey(0), (65536,), dtype=jnp.float16)

# Forward pass: correct and finite
y = jax.scipy.special.log_softmax(x)
print(jnp.isfinite(y).all())   # True

# Gradient: silently wrong
g = jax.grad(lambda x: jax.scipy.special.log_softmax(x).sum())(x)
print(jnp.isfinite(g).all())     # False
print((~jnp.isfinite(g)).sum())  # 65536 -- every element is NaN

# Also affects jax.nn.log_softmax, and JIT:
g2 = jax.jit(jax.grad(lambda x: jax.nn.log_softmax(x, axis=-1).sum()))(x)
print(jnp.isfinite(g2).all())  # False

# True gradient is small, well within float16's range:
g32 = jax.grad(lambda x: jax.scipy.special.log_softmax(x).sum())(x.astype(jnp.float32))
print(float(g32.min()), float(g32.max()))   # ~-13.66, ~0.98

# Contrast: a linear reduction whose true gradient IS out of range
# saturates to -inf, not NaN, as expected:
g3 = jax.grad(lambda x: (x - jnp.sum(x)).sum())(x)
print(jnp.isneginf(g3).all())  # True

# Ruling out max-tie smearing (cf. #13529): force a single, unambiguous max
x_unique_max = x.at[0].set(jnp.float16(6.0))
g4 = jax.grad(lambda x: jax.scipy.special.log_softmax(x).sum())(x_unique_max)
print(jnp.isfinite(g4).all())  # Still False
```

### Error

No exception is raised — NaN is returned silently.

### Expected Behavior

The gradient should be finite for a well-conditioned finite input where the true gradient is small, matching float32 behavior on the same data.

### System info (python version, jaxlib version, accelerator, etc.)

jax:    0.10.1
jaxlib: 0.10.1
numpy:  2.4.6
python: 3.12.13 | packaged by Anaconda, Inc. | (main, Mar 19 2026, 20:20:58) [GCC 14.3.0]
device info: cpu-1, 1 local devices
process_count: 1
platform: uname_result(system='Linux', release='6.8.0-136-generic', machine='x86_64')
JAX_PLATFORMS=cpu

### PR DESCRIPTION
**Pull request by YusefSyed**
Fixes #40409.

For float16 inputs, `log_softmax` can overflow its reverse-mode cotangent reduction even when the resulting gradients are representable in float16. Perform the internal computation in float32 and cast the output back to float16 so the backward terms combine before the final cast. Other input dtypes retain their existing path.

The regression tests cover large masked and unmasked inputs, multiple axis forms, eager/JIT VJPs, and cancellation where an intermediate term exceeds float16 range but the final gradient fits. All 14 new cases fail against the unchanged production baseline and pass with this change.

Validation on Apple Silicon CPU:

- `pytest tests/nn_test.py -q`: 404 passed, 19 skipped.
- `JAX_ENABLE_X64=True pytest tests/nn_test.py -q`: 452 passed, 19 skipped.
- Applicable pre-commit hooks, including Ruff and Pyrefly: passed.

GPU/TPU performance and backward-residual memory usage have not been benchmarked. Float16 rounding can change because the intermediate computation uses float32.

AI assistance: Codex assisted with investigation, implementation, testing, and this description. The account holder approved the patch and authorized submission. The listed commands were executed through Codex.

### CODE REVIEW

**Comment by YusefSyed** [comment]
Following up on this fix for float16 `log_softmax` gradient overflow. The patch widens only the internal backward computation to float32, then returns float16 as before. Could someone familiar with this path review whether that dtype boundary is appropriate? Happy to adjust the approach.
